test/test_base64.c

```diff
@@ -2,15 +2,79 @@
 #include <stdio.h>
 #include "../include/libbase64.h"
 #include "codec_supported.h"
+#include <malloc.h>
 
 static int fail = 0;
 static char out[2000];
 static size_t outlen;
 
-extern char _binary_moby_dick_plain_txt_start[];
-extern char _binary_moby_dick_plain_txt_end[];
-extern char _binary_moby_dick_base64_txt_start[];
-extern char _binary_moby_dick_base64_txt_end[];
+#ifdef _MSC_VER
+	char* read_all(const char* file, long* length) {
+		FILE    *infile;
+		char    *buffer;
+		long    numbytes;
+
+		/* open an existing file for reading */
+		infile = fopen(file, "rb");
+
+		/* quit if the file does not exist */
+		if (infile == NULL)
+			return 0;
+
+		/* Get the number of bytes */
+		fseek(infile, 0L, SEEK_END);
+		numbytes = ftell(infile);
+		*length = numbytes;
+
+		/* reset the file position indicator to
+		the beginning of the file */
+		fseek(infile, 0L, SEEK_SET);
+
+		/* grab sufficient memory for the
+		buffer to hold the text */
+		buffer = (char*)malloc(numbytes+1);
+
+		/* memory error */
+		if (buffer == NULL)
+			return 0;
+
+		/* copy all the text into the buffer */
+		fread(buffer, sizeof(char), numbytes, infile);
+		fclose(infile);
+		buffer[numbytes] = 0;
+		
+		/* free the memory we used for the buffer */
+		return buffer;
+	}
+
+#define READ_INPUT_FILE(filename, name) \
+			{\
+		long length; \
+		char* data ;\
+		data = read_all(filename, &length); \
+		if ( !data ) {\
+			printf("FAIL: could not read input file %s", filename);\
+			return 1;\
+		}\
+		if (strstr(data, "\r\n") != NULL) {\
+			printf("FAIL: Input file %s must have UNIX-style line endings", filename);\
+			return 1;\
+		}\
+		_binary_##name##_start = data;\
+		_binary_##name##_end = data+length;\
+	}
+	char* _binary_moby_dick_plain_txt_start;
+	char* _binary_moby_dick_plain_txt_end;
+	char* _binary_moby_dick_base64_txt_start;
+	char* _binary_moby_dick_base64_txt_end;
+	
+#else
+	extern char _binary_moby_dick_plain_txt_start[];
+	extern char _binary_moby_dick_plain_txt_end[];
+	extern char _binary_moby_dick_base64_txt_start[];
+	extern char _binary_moby_dick_base64_txt_end[];
+#endif
+
```

Comment:
For reading arbitrary files, there's already the `base64` binary in /bin :)  For testing I like the idea of a self-contained binary that contains both the input and the expected output. If you use arbitrary files, where do you get the validated, expected output from?

To fix this issue, I propose to include the Moby Dick passages as a char array in a header file. That should be cross-platform compatible. I will make that change, probably tomorrow at soonest.